packages/gatsby-script/src/gatsby-script.tsx

```diff
@@ -0,0 +1,158 @@
+import React, { useEffect } from "react"
+import type { ReactElement, ScriptHTMLAttributes } from "react"
+
+export enum ScriptStrategy {
+  preHydrate = `pre-hydrate`,
+  postHydrate = `post-hydrate`,
+  idle = `idle`,
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ScriptProps
+  extends Omit<ScriptHTMLAttributes<HTMLScriptElement>, `onLoad` | `onError`> {
+  id?: string
+  strategy?: ScriptStrategy
+  children?: string
+  onLoad?: (event: Event) => void
+  onError?: (event: ErrorEvent) => void
+}
+
+const handledProps = new Set([
+  `src`,
+  `strategy`,
+  `dangerouslySetInnerHTML`,
+  `children`,
+  `onLoad`,
+  `onError`,
+])
+
+export const scriptCache = new Set()
+
+export function Script(props: ScriptProps): ReactElement {
+  const {
+    src,
+    strategy = ScriptStrategy.postHydrate,
+    onLoad,
+    onError,
+  } = props || {}
+
+  useEffect(() => {
+    let script: HTMLScriptElement | null
+
+    switch (strategy) {
+      case ScriptStrategy.preHydrate:
+        // If the navigation is client-side (e.g. via gatsby-link), treat it like post-hydrate.
+        // We probably want to make the router write to history so we can compare that instead.
+        // The current approach doesn't solve the browser back/forward button use case.
+        if (!performance.getEntriesByName(location.href).length) {
```

Comment:
This is smart way to look for scripts already being loaded, the problem here is that there is a limit to how many resources the browser keeps in its buffer. By default, it's around 100-150 entries so it's better to use a Map instead.